internal/dataplane/translator/translator.go

```diff
@@ -592,8 +592,17 @@ func (t *Translator) getGatewayListeningPorts(
 		// If no sectionName is specified, all ports are used (according to the specification
 		// "When unspecified (empty string), this will reference the entire resource." - see
 		// https://github.com/kubernetes-sigs/gateway-api/blob/ebe9f31ef27819c3b29f698a3e9b91d279453c59/apis/v1/shared_types.go#L107).
+		//
+		// When the parent resource is a Gateway, this targets all listeners
+		// listening on the specified port that also support this kind of Route(and
+		// select this Route). It's not recommended to set `Port` unless the
+		// networking behaviors specified in a Route must apply to a specific port
+		// as opposed to a listener(s) whose port(s) may be changed. When both Port
+		// and SectionName are specified, the name and port of the selected listener
+		// must match both specified values. - see
+		// https://github.com/kubernetes-sigs/gateway-api/blob/e2f02aa0fc9dfa27be271397801f52f4e0a78acc/apis/v1/shared_types.go#L124C1-L130C38
 		gwPorts = append(gwPorts, lo.FilterMap(gw.Spec.Listeners, func(l gatewayapi.Listener, _ int) (gatewayapi.PortNumber, bool) {
-			if (pr.SectionName == nil || *pr.SectionName == l.Name) && protocol == l.Protocol {
+			if (pr.SectionName == nil || (*pr.SectionName == l.Name && pr.Port == nil) || (*pr.SectionName == l.Name && *pr.Port == l.Port)) && protocol == l.Protocol {
```

Comment:
Since `the name and port of the selected listener must match both specified values` I wonder if it'd be beneficial to use `t.registerTranslationFailure(...)` in case of such a mismatch to present it to the user 🤔 

Whar are we doing in other places?

Reply:
good point :+1: 

leave a log can provide clearer prompts for the user, or should we update it to the status?

Reply:
Discussed in the team sync
We should disallow the configuration with a port/section mismatch and:
1. (ideally) put the error in the status of the object
2. ~emit an `Event`~ rejected because we only emit events in a later stage of config processing
3. ~(possibly) also log - but only if 1 and 2 are not possible~ - rejected because `Event`s are the way to report translation errors

Reply:
@programmer04 I have created a follow-up issue for tracking purposes and will address it in future pull requests.

https://github.com/Kong/kubernetes-ingress-controller/issues/5157

Let's keep this PR simple. What do you think?

Reply:
I think that makes sense to tackle this separately. 👍 

Regarding https://github.com/Kong/kubernetes-ingress-controller/issues/5157: can we please make it precise in terms of what we're going to implement? Now it's very general while I think it should be specific to TCPRoute/UDPRoute/HTTPRoute and their attachment to Gateway listeners. As I understand, we want to make their attachment compliant with the [Gateway API spec](https://github.com/kubernetes-sigs/gateway-api/blob/e2f02aa0fc9dfa27be271397801f52f4e0a78acc/apis/v1/shared_types.go#L142C16-L142C16) - they shouldn't be attached to a Gateway when there's no matching port in Gateway for them.

> 	 For the purpose of status, an attachment is considered successful as
	 long as the parent resource accepts it partially. For example, Gateway
	 listeners can restrict which Routes can attach to them by Route kind,
	 namespace, or hostname. If 1 of 2 Gateway listeners accept attachment
	 from the referencing Route, the Route MUST be considered successfully
	 attached. If no Gateway listeners accept attachment from this Route,
	 the Route MUST be considered detached from the Gateway.

BTW it seems to me it may be already implemented as expected, but in an earlier stage of processing i.e.[in the controller](https://github.com/Kong/kubernetes-ingress-controller/blob/dce791d4e434047355570bf12ef1f14a2a685d98/internal/controllers/gateway/route_utils.go#L311). That means that if a Route doesn't have a matching Gateway, it will never get to the stage we're modifying in this PR (translation) because it will never get to [UpdateObject in the controller](https://github.com/Kong/kubernetes-ingress-controller/blob/2edf55e7cade5cb042f80e2b1fe5d95ea36ce663/internal/controllers/gateway/httproute_controller.go#L419) (will be rejected [here](https://github.com/Kong/kubernetes-ingress-controller/blob/2edf55e7cade5cb042f80e2b1fe5d95ea36ce663/internal/controllers/gateway/httproute_controller.go#L373) before).

IMO a reasonable way to approach this would be to ensure there's a Gateway API conformance test for this part of spec first. If there's not, add it in the upstream first (as I can see it's in the process: https://github.com/kubernetes-sigs/gateway-api/pull/2582). #5157 would be then an issue that will make sure we pass this new/already existing test.

Reply:
Thanks! @czeslavo 

I will update the description of #51557

I will conduct in-depth research on the relevant logic you mentioned. In addition, perhaps we need to wait for the merger of https://github.com/kubernetes-sigs/gateway-api/pull/2582

Reply:
> IMO a reasonable way to approach this would be to ensure there's a Gateway API conformance test for this part of spec first. If there's not, add it in the upstream first (as I can see it's in the process: https://github.com/kubernetes-sigs/gateway-api/pull/2582). https://github.com/Kong/kubernetes-ingress-controller/issues/5157 would be then an issue that will make sure we pass this new/already existing test.

I agree that having conformance tests about it is the best way to do it, and I really liked the  https://github.com/kubernetes-sigs/gateway-api/pull/2582 effort to test it for `HTTPRoutes`. Unfortunately, At the moment, no `TCPRoute`/`UDPRoute` tests exist, and I do not think there is will to create them as they are still alpha and there is uncertainty about their future.

Reply:
I think we may need to improve [this portion of code](https://github.com/Kong/kubernetes-ingress-controller/blob/2edf55e7cade5cb042f80e2b1fe5d95ea36ce663/internal/controllers/gateway/route_utils.go#L233-L256) with the check on section and port and the subsequent status update.

Reply:
I have added a `blocked` label,  after https://github.com/kubernetes-sigs/gateway-api/pull/2582 got merged, I will continue this PR.